proxy.go

```diff
@@ -51,108 +51,86 @@ func (ph ProxyHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 }
 
 func (ph ProxyHandler) handleConnect(w http.ResponseWriter, req *http.Request) {
-	h, ok := w.(http.Hijacker)
-	if !ok {
-		msg := fmt.Sprintf("Can't hijack connection to %v", req.Host)
-		http.Error(w, msg, http.StatusInternalServerError)
-		return
-	}
 	u, err := ph.transport.Proxy(req)
 	if err != nil {
 		http.Error(w, err.Error(), http.StatusInternalServerError)
 		return
 	}
 	var server net.Conn
 	if u == nil {
-		server = connectToServer(w, req)
+		server, err = net.Dial("tcp", req.Host)
 	} else {
-		server = connectViaProxy(w, req, u.Host, ph.auth)
+		server, err = connectViaProxy(req, u.Host, ph.auth)
 	}
-	if server == nil {
+	if err != nil {
+		w.WriteHeader(http.StatusBadGateway)
+		return
+	}
+	h, ok := w.(http.Hijacker)
+	if !ok {
+		msg := fmt.Sprintf("Can't hijack connection to %v", req.Host)
```

Comment:
This message is a little strange. You can't hijack the connection because it is not hijackable. It doesn't really have anything to do with the `req.Host` that you want to connect to. This is more of an internal error, but the message suggests some sort of external cause. And if you are going via a proxy, the connection isn't really "to `req.Host`" either.

Reply:
Yes, you're right. Technically it's a connection going back to `req.RemoteAddr`. If I changed it to say that, do you think that's enough? Or are you also saying that it's just not a very user friendly error message?

Reply:
Also it's not really hijacking a connection, it's hijacking a response (writer). So how about the following?

`fmt.Sprintf("Can't hijack response back to %v", req.RemoteAddr)`

Reply:
I guess it doesn't really matter what it says - there's not much you can do if you get the error - I'm not sure using `req.RemoteAddr` makes much difference either. The original message just seemed odd when I read it, but I guess `req.Host` is the most specific part of the connection so probably identifies it best even if not technically correct. You can leave it as is if you want.

Reply:
Looking at this a bit closer, I feel like these error messages shouldn't get sent back in a 500 response body anyway - they should be logged together with the request id. We probably don't even need to send a response body back at all, so this could just change to `w.WriteHeader(http.StatusInternalServerError)`.

There are a number of places that this happens elsewhere in the code, so I'll go through and make the changes in a separate pull request.